File: packages/parent-domain/src/parent-assistant.ts
import { type Infer, Schema, withParser } from '@ocentra-parent/schema-domain/effect';
import {
  LocalAiDegradedStateSchema,
  LocalAiModelIdSchema,
  LocalAiProviderIdSchema,
  LocalAiResultIdSchema,
  LocalAiTimestampSchema,
  LocalAiUnavailableReasonSchema,
} from './local-ai-primitives';
import { LocalAiProviderSchedulerJobStatusSchema } from './local-ai-provider-scheduler';
import {
  FamilyReferenceSchema,
  ParentActionReferenceSchema,
  ParentActorReferenceSchema,
  ParentDeviceReferenceSchema,
  ParentEvidenceReferenceSchema,
} from './references';
import { ParentContractSchemaVersionSchema } from './reference-primitives';

const NonEmptyParentAssistantText = Schema.String.pipe(Schema.minLength(1));
const ParentAssistantPositiveCount = Schema.Number.pipe(Schema.int(), Schema.positive());

export const ParentAssistantRequestIdSchema = NonEmptyParentAssistantText.pipe(
  Schema.brand('ParentAssistantRequestId')
);
export const ParentAssistantThreadIdSchema = NonEmptyParentAssistantText.pipe(Schema.brand('ParentAssistantThreadId'));
export const ParentAssistantMessageIdSchema = NonEmptyParentAssistantText.pipe(
  Schema.brand('ParentAssistantMessageId')
);
export const ParentAssistantPromptVersionSchema = NonEmptyParentAssistantText.pipe(
  Schema.brand('ParentAssistantPromptVersion')
);
export const ParentAssistantQuestionSchema = NonEmptyParentAssistantText.pipe(Schema.brand('ParentAssistantQuestion'));
export const ParentAssistantApiProviderIdSchema = NonEmptyParentAssistantText.pipe(
  Schema.brand('ParentAssistantApiProviderId')
);
export const ParentAssistantCustodyLabelSchema = NonEmptyParentAssistantText.pipe(
  Schema.brand('ParentAssistantCustodyLabel')
);
export const ParentAssistantRetentionPolicySchema = NonEmptyParentAssistantText.pipe(
  Schema.brand('ParentAssistantRetentionPolicy')
);
export const ParentAssistantDeletionPolicySchema = NonEmptyParentAssistantText.pipe(
  Schema.brand('ParentAssistantDeletionPolicy')
);
export const ParentAssistantAnswerTextSchema = NonEmptyParentAssistantText.pipe(
  Schema.brand('ParentAssistantAnswerText')
);
export const ParentAssistantCitationLabelSchema = NonEmptyParentAssistantText.pipe(
  Schema.brand('ParentAssistantCitationLabel')
);
export const ParentAssistantActionPreviewIdSchema = NonEmptyParentAssistantText.pipe(
  Schema.brand('ParentAssistantActionPreviewId')
);
export const ParentAssistantRunIdSchema = NonEmptyParentAssistantText.pipe(Schema.brand('ParentAssistantRunId'));
export const ParentAssistantActionIntentIdSchema = NonEmptyParentAssistantText.pipe(
  Schema.brand('ParentAssistantActionIntentId')
);

export const ParentAssistantBackendStateSchema = withParser(
  Schema.Literal('runtime-backed', 'volatile-local', 'contract-required', 'unavailable')
);
export const ParentAssistantProviderStateSchema = withParser(Schema.Literal('configured', 'degraded', 'unavailable'));
export const ParentAssistantAnswerStateSchema = withParser(
  Schema.Literal('answered', 'queued', 'degraded', 'unavailable')
);
export const ParentAssistantApiAuthorizationStateSchema = withParser(Schema.Literal('authorized', 'not-authorized'));
export const ParentAssistantActionPreviewKindSchema = withParser(
  Schema.Literal('none', 'policy-suggestion', 'schedule-change', 'time-limit-change')
);
export const ParentAssistantThreadStateSchema = withParser(Schema.Literal('open', 'archived'));
export const ParentAssistantRunCancelStateSchema = withParser(
  Schema.Literal('cancelled', 'not-running', 'unavailable')
);
export const ParentAssistantActionConfirmStateSchema = withParser(
  Schema.Literal('contract-required', 'not-applied', 'rejected')
);

export const ParentAssistantScopeSchema = withParser(
  Schema.Struct({
    family: FamilyReferenceSchema,
    device: Schema.Union(ParentDeviceReferenceSchema, Schema.Null),
  })
);

export const ParentAssistantEvidenceContextSchema = withParser(
  Schema.Struct({
    evidence: ParentEvidenceReferenceSchema,
    citationLabel: ParentAssistantCitationLabelSchema,
    allowedSummary: ParentAssistantAnswerTextSchema,
  })
);

export const ParentAssistantActionPreviewSchema = withParser(
  Schema.Struct({
    previewId: Schema.Union(ParentAssistantActionPreviewIdSchema, Schema.Null),
    actionKind: ParentAssistantActionPreviewKindSchema,
    summary: Schema.Union(ParentAssistantAnswerTextSchema, Schema.Null),
    actionReference: Schema.Union(ParentActionReferenceSchema, Schema.Null),
    requiresControllerLease: Schema.Boolean,
    childAgentContractRequired: Schema.Literal(true),
    enforcementApplied: Schema.Literal(false),
  })
);

export const ParentAssistantGenerateRequestSchema = withParser(
  Schema.Struct({
    schemaVersion: ParentContractSchemaVersionSchema,
    requestId: ParentAssistantRequestIdSchema,
    threadId: ParentAssistantThreadIdSchema,
    messageId: ParentAssistantMessageIdSchema,
    askedAt: LocalAiTimestampSchema,
    actor: ParentActorReferenceSchema,
    scope: ParentAssistantScopeSchema,
    question: ParentAssistantQuestionSchema,
    evidenceContext: Schema.Array(ParentAssistantEvidenceContextSchema),
    modelId: Schema.Union(LocalAiModelIdSchema, Schema.Null),
    maxOutputTokens: ParentAssistantPositiveCount,
    timeoutMs: ParentAssistantPositiveCount,
  })
);

const ParentAssistantApiProviderBoundaryBaseSchema = Schema.Struct({
  schemaVersion: ParentContractSchemaVersionSchema,
  providerId: ParentAssistantApiProviderIdSchema,
  authorizationState: ParentAssistantApiAuthorizationStateSchema,
  custodyLabel: ParentAssistantCustodyLabelSchema,
  retentionPolicy: ParentAssistantRetentionPolicySchema,
  deletionPolicy: ParentAssistantDeletionPolicySchema,
  citations: Schema.Array(ParentAssistantEvidenceContextSchema),
  providerState: ParentAssistantProviderStateSchema,
  unavailableReason: Schema.Union(LocalAiUnavailableReasonSchema, Schema.Null),
  childSafetyOrEnforcementUseAllowed: Schema.Literal(false),
});

type ParentAssistantApiProviderBoundaryCandidate = Infer<typeof ParentAssistantApiProviderBoundaryBaseSchema>;

export const ParentAssistantApiProviderBoundarySchema = withParser(
  ParentAssistantApiProviderBoundaryBaseSchema.pipe(
    Schema.filter(
      (boundary) =>
        parentAssistantApiProviderBoundaryIsConsistent(boundary) ||
        'Expected API AI provider use to require parent authorization, citations, custody, retention, deletion, and no child-safety enforcement use'
    )
  )
);

const ParentAssistantAnswerBaseSchema = Schema.Struct({
  schemaVersion: ParentContractSchemaVersionSchema,
  requestId: ParentAssistantRequestIdSchema,
  threadId: ParentAssistantThreadIdSchema,
  messageId: ParentAssistantMessageIdSchema,
  answeredAt: LocalAiTimestampSchema,
  providerId: LocalAiProviderIdSchema,
  modelId: LocalAiModelIdSchema,
  providerState: ParentAssistantProviderStateSchema,
  answerState: ParentAssistantAnswerStateSchema,
  schedulerJobStatus: LocalAiProviderSchedulerJobStatusSchema,
  degradedState: LocalAiDegradedStateSchema,
  unavailableReason: Schema.Union(LocalAiUnavailableReasonSchema, Schema.Null),
  localAiResultId: Schema.Union(LocalAiResultIdSchema, Schema.Null),
  answerText: Schema.Union(ParentAssistantAnswerTextSchema, Schema.Null),
  citations: Schema.Array(ParentAssistantEvidenceContextSchema),
  actionPreview: ParentAssistantActionPreviewSchema,
  apiProviderBoundary: ParentAssistantApiProviderBoundarySchema,
  promptVersion: ParentAssistantPromptVersionSchema,
});

type ParentAssistantAnswerCandidate = Infer<typeof ParentAssistantAnswerBaseSchema>;

export const ParentAssistantAnswerSchema = withParser(
  ParentAssistantAnswerBaseSchema.pipe(
    Schema.filter(
      (answer) =>
        parentAssistantAnswerIsConsistent(answer) ||
        'Expected parent assistant answers to cite evidence when answered and expose unavailable reason when unavailable'
    )
  )
);

export const ParentAssistantThreadRecordSchema = withParser(
  Schema.Struct({
    schemaVersion: ParentContractSchemaVersionSchema,
    threadId: ParentAssistantThreadIdSchema,
    title: ParentAssistantAnswerTextSchema,
    state: ParentAssistantThreadStateSchema,
    backendState: ParentAssistantBackendStateSchema,
    createdAt: LocalAiTimestampSchema,
    updatedAt: LocalAiTimestampSchema,
    messageCount: Schema.Number.pipe(Schema.int()),
  })
);

export const ParentAssistantThreadResponseSchema = withParser(
  Schema.Struct({
    schemaVersion: ParentContractSchemaVersionSchema,
    backendState: ParentAssistantBackendStateSchema,
    activeThread: Schema.Union(ParentAssistantThreadRecordSchema, Schema.Null),
    threads: Schema.Array(ParentAssistantThreadRecordSchema),
    reason: Schema.Union(ParentAssistantAnswerTextSchema, Schema.Null),
  })
);

export const ParentAssistantProviderStatusSchema = withParser(
  Schema.Struct({
    schemaVersion: ParentContractSchemaVersionSchema,
    backendState: ParentAssistantBackendStateSchema,
    providerId: LocalAiProviderIdSchema,
    modelId: LocalAiModelIdSchema,
    providerState: ParentAssistantProviderStateSchema,
    schedulerJobStatus: LocalAiProviderSchedulerJobStatusSchema,
    degradedState: LocalAiDegradedStateSchema,
    unavailableReason: Schema.Union(LocalAiUnavailableReasonSchema, Schema.Null),
    queueDepth: Schema.Number.pipe(Schema.int()),
    busy: Schema.Boolean,
    apiProviderBoundary: ParentAssistantApiProviderBoundarySchema,
  })
);

export const ParentAssistantRunCancelResultSchema = withParser(
  Schema.Struct({
    schemaVersion: ParentContractSchemaVersionSchema,
    backendState: ParentAssistantBackendStateSchema,
    threadId: ParentAssistantThreadIdSchema,
    runId: ParentAssistantRunIdSchema,
    cancelState: ParentAssistantRunCancelStateSchema,
    providerState: ParentAssistantProviderStateSchema,
    unavailableReason: Schema.Union(LocalAiUnavailableReasonSchema, Schema.Null),
  })
);

const ParentAssistantActionConfirmResultBaseSchema = Schema.Struct({
  schemaVersion: ParentContractSchemaVersionSchema,
  backendState: ParentAssistantBackendStateSchema,
  actionIntentId: ParentAssistantActionIntentIdSchema,
  previewId: Schema.Union(ParentAssistantActionPreviewIdSchema, Schema.Null),
  actionKind: ParentAssistantActionPreviewKindSchema,
  confirmState: ParentAssistantActionConfirmStateSchema,
  requiresControllerLease: Schema.Literal(true),
  childAgentContractRequired: Schema.Literal(true),
  enforcementApplied: Schema.Literal(false),
  policyWritten: Schema.Literal(false),
  reason: ParentAssistantAnswerTextSchema,
});

type ParentAssistantActionConfirmResultCandidate = Infer<typeof ParentAssistantActionConfirmResultBaseSchema>;

export const ParentAssistantActionConfirmResultSchema = withParser(
  ParentAssistantActionConfirmResultBaseSchema.pipe(
    Schema.filter(
      (result) =>
        parentAssistantActionConfirmResultIsSafe(result) ||
        'Expected Parent Assistant action confirm to require controller/child-agent contract and avoid direct enforcement or policy writes'
    )
  )
);

function parentAssistantAnswerIsConsistent(answer: ParentAssistantAnswerCandidate): boolean {
  if (answer.answerState === 'answered') {
    return (
      answer.answerText !== null &&
      answer.citations.length > 0 &&
      answer.unavailableReason === null &&
      answer.providerState === 'configured'
    );
  }

  if (answer.answerState === 'unavailable') {
    return answer.answerText === null && answer.unavailableReason !== null && answer.providerState === 'unavailable';
  }

  if (answer.answerState === 'degraded') {
    return answer.degradedState !== 'none' && answer.providerState === 'degraded';
  }

  return answer.schedulerJobStatus === 'queued' && answer.answerText === null;
}

function parentAssistantApiProviderBoundaryIsConsistent(
  boundary: ParentAssistantApiProviderBoundaryCandidate
): boolean {
  if (boundary.citations.length === 0 || boundary.childSafetyOrEnforcementUseAllowed !== false) {
    return false;
  }

  if (boundary.authorizationState === 'not-authorized') {
    return boundary.providerState === 'unavailable' && boundary.unavailableReason !== null;
  }

  return boundary.providerState !== 'unavailable' || boundary.unavailableReason !== null;
}

function parentAssistantActionConfirmResultIsSafe(result: ParentAssistantActionConfirmResultCandidate): boolean {
  return (
    result.requiresControllerLease === true &&
    result.childAgentContractRequired === true &&
    result.enforcementApplied === false &&
    result.policyWritten === false &&
    result.confirmState === 'contract-required'
  );
}

export type ParentAssistantScope = Infer<typeof ParentAssistantScopeSchema>;
export type ParentAssistantEvidenceContext = Infer<typeof ParentAssistantEvidenceContextSchema>;
export type ParentAssistantActionPreview = Infer<typeof ParentAssistantActionPreviewSchema>;
export type ParentAssistantGenerateRequest = Infer<typeof ParentAssistantGenerateRequestSchema>;
export type ParentAssistantAnswer = Infer<typeof ParentAssistantAnswerSchema>;
export type ParentAssistantApiProviderBoundary = Infer<typeof ParentAssistantApiProviderBoundarySchema>;
export type ParentAssistantThreadRecord = Infer<typeof ParentAssistantThreadRecordSchema>;
export type ParentAssistantThreadResponse = Infer<typeof ParentAssistantThreadResponseSchema>;
export type ParentAssistantProviderStatus = Infer<typeof ParentAssistantProviderStatusSchema>;
export type ParentAssistantRunCancelResult = Infer<typeof ParentAssistantRunCancelResultSchema>;
export type ParentAssistantActionConfirmResult = Infer<typeof ParentAssistantActionConfirmResultSchema>;

File: crates/agent-protocol/src/parent_assistant.rs
use serde::{Deserialize, Serialize};

use crate::{
    FamilyReference, LocalAiDegradedState, LocalAiProviderSchedulerJobStatus,
    LocalAiProviderSchedulerStatus, ParentActionReference, ParentActorReference,
    ParentDeviceReference, ParentEvidenceReference,
};

pub mod provider_route;

use provider_route::ParentAssistantProviderRoute;

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub enum ParentAssistantProviderState {
    #[serde(rename = "configured")]
    Configured,
    #[serde(rename = "degraded")]
    Degraded,
    #[serde(rename = "unavailable")]
    Unavailable,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub enum ParentAssistantAnswerState {
    #[serde(rename = "answered")]
    Answered,
    #[serde(rename = "queued")]
    Queued,
    #[serde(rename = "degraded")]
    Degraded,
    #[serde(rename = "unavailable")]
    Unavailable,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub enum ParentAssistantRunState {
    #[serde(rename = "queued")]
    Queued,
    #[serde(rename = "active")]
    Active,
    #[serde(rename = "completed")]
    Completed,
    #[serde(rename = "failed")]
    Failed,
    #[serde(rename = "cancelled")]
    Cancelled,
    #[serde(rename = "degraded")]
    Degraded,
    #[serde(rename = "unavailable")]
    Unavailable,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub enum ParentAssistantApiAuthorizationState {
    #[serde(rename = "authorized")]
    Authorized,
    #[serde(rename = "not-authorized")]
    NotAuthorized,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub enum ParentAssistantApiProviderAccessState {
    #[serde(rename = "not-authorized")]
    NotAuthorized,
    #[serde(rename = "authorized-unavailable")]
    AuthorizedUnavailable,
    #[serde(rename = "authorized-degraded")]
    AuthorizedDegraded,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub enum ParentAssistantActionPreviewKind {
    #[serde(rename = "none")]
    None,
    #[serde(rename = "policy-suggestion")]
    PolicySuggestion,
    #[serde(rename = "schedule-change")]
    ScheduleChange,
    #[serde(rename = "time-limit-change")]
    TimeLimitChange,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub enum ParentAssistantActionPreviewState {
    #[serde(rename = "draft")]
    Draft,
    #[serde(rename = "unavailable")]
    Unavailable,
    #[serde(rename = "rejected")]
    Rejected,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub enum ParentAssistantBackendState {
    #[serde(rename = "runtime-backed")]
    RuntimeBacked,
    #[serde(rename = "durable-local")]
    DurableLocal,
    #[serde(rename = "volatile-local")]
    VolatileLocal,
    #[serde(rename = "contract-required")]
    ContractRequired,
    #[serde(rename = "unavailable")]
    Unavailable,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub enum ParentAssistantThreadState {
    #[serde(rename = "open")]
    Open,
    #[serde(rename = "archived")]
    Archived,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub enum ParentAssistantRunCancelState {
    #[serde(rename = "cancelled")]
    Cancelled,
    #[serde(rename = "not-running")]
    NotRunning,
    #[serde(rename = "unavailable")]
    Unavailable,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub enum ParentAssistantActionConfirmState {
    #[serde(rename = "contract-required")]
    ContractRequired,
    #[serde(rename = "not-applied")]
    NotApplied,
    #[serde(rename = "rejected")]
    Rejected,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub enum ParentAssistantChildAgentValidationState {
    #[serde(rename = "child-agent-contract-required")]
    ChildAgentContractRequired,
    #[serde(rename = "child-agent-offline")]
    ChildAgentOffline,
    #[serde(rename = "child-agent-unavailable")]
    ChildAgentUnavailable,
    #[serde(rename = "child-agent-degraded")]
    ChildAgentDegraded,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ParentAssistantScope {
    pub family: FamilyReference,
    pub device: Option<ParentDeviceReference>,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ParentAssistantEvidenceContext {
    pub evidence: ParentEvidenceReference,
    pub citation_label: String,
    pub allowed_summary: String,
    pub custody_label: String,
    pub source_label: String,
    pub raw_child_evidence_included: bool,
    pub direct_enforcement_allowed: bool,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ParentAssistantActionPreview {
    pub preview_id: Option<String>,
    pub action_kind: ParentAssistantActionPreviewKind,
    pub summary: Option<String>,
    pub action_reference: Option<ParentActionReference>,
    pub requires_controller_lease: bool,
    pub child_agent_contract_required: bool,
    pub enforcement_applied: bool,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ParentAssistantActionPreviewResult {
    pub schema_version: String,
    pub backend_state: ParentAssistantBackendState,
    pub action_intent_id: String,
    pub preview_state: ParentAssistantActionPreviewState,
    pub preview: ParentAssistantActionPreview,
    pub evidence_context: Vec<ParentAssistantEvidenceContext>,
    pub preview_required: bool,
    pub preview_satisfied: bool,
    pub raw_assistant_prose_accepted: bool,
    pub parent_confirmation_required: bool,
    pub parent_confirmation_recorded: bool,
    pub child_agent_validation_state: ParentAssistantChildAgentValidationState,
    pub source_refs: Vec<ParentEvidenceReference>,
    pub audit_reason: String,
    pub requires_controller_lease: bool,
    pub child_agent_contract_required: bool,
    pub enforcement_applied: bool,
    pub policy_written: bool,
    pub reason: String,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ParentAssistantGenerateRequest {
    pub schema_version: String,
    pub request_id: String,
    pub thread_id: String,
    pub message_id: String,
    pub asked_at: String,
    pub actor: ParentActorReference,
    pub scope: ParentAssistantScope,
    pub question: String,
    pub evidence_context: Vec<ParentAssistantEvidenceContext>,
    pub model_id: Option<String>,
    pub max_output_tokens: u32,
    pub timeout_ms: u64,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ParentAssistantApiProviderBoundary {
    pub schema_version: String,
    pub provider_id: String,
    pub authorization_state: ParentAssistantApiAuthorizationState,
    pub access_state: ParentAssistantApiProviderAccessState,
    pub parent_authorization_required: bool,
    pub evidence_citation_required: bool,
    pub custody_label: String,
    pub custody_state: String,
    pub retention_policy: String,
    pub retention_state: String,
    pub deletion_policy: String,
    pub deletion_state: String,
    pub citations: Vec<ParentAssistantEvidenceContext>,
    pub provider_state: ParentAssistantProviderState,
    pub unavailable_reason: Option<String>,
    pub child_safety_or_enforcement_use_allowed: bool,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ParentAssistantAnswer {
    pub schema_version: String,
    pub request_id: String,
    pub thread_id: String,
    pub message_id: String,
    pub answered_at: String,
    pub provider_id: String,
    pub model_id: String,
    pub provider_state: ParentAssistantProviderState,
    pub answer_state: ParentAssistantAnswerState,
    pub run_state: ParentAssistantRunState,
    pub scheduler_job_status: LocalAiProviderSchedulerJobStatus,
    pub degraded_state: LocalAiDegradedState,
    pub unavailable_reason: Option<String>,
    pub local_ai_result_id: Option<String>,
    pub answer_text: Option<String>,
    pub citations: Vec<ParentAssistantEvidenceContext>,
    pub action_preview: ParentAssistantActionPreview,
    pub api_provider_boundary: ParentAssistantApiProviderBoundary,
    pub provider_route: ParentAssistantProviderRoute,
    pub prompt_version: String,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ParentAssistantThreadRecord {
    pub schema_version: String,
    pub thread_id: String,
    pub title: String,
    pub state: ParentAssistantThreadState,
    pub backend_state: ParentAssistantBackendState,
    pub created_at: String,
    pub updated_at: String,
    pub message_count: u32,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ParentAssistantThreadResponse {
    pub schema_version: String,
    pub backend_state: ParentAssistantBackendState,
    pub active_thread: Option<ParentAssistantThreadRecord>,
    pub threads: Vec<ParentAssistantThreadRecord>,
    pub reason: Option<String>,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ParentAssistantProviderStatus {
    pub schema_version: String,
    pub backend_state: ParentAssistantBackendState,
    pub provider_id: String,
    pub model_id: String,
    pub provider_state: ParentAssistantProviderState,
    pub run_state: ParentAssistantRunState,
    pub scheduler_job_status: LocalAiProviderSchedulerJobStatus,
    pub scheduler_status: LocalAiProviderSchedulerStatus,
    pub degraded_state: LocalAiDegradedState,
    pub unavailable_reason: Option<String>,
    pub queue_depth: u16,
    pub busy: bool,
    pub api_provider_boundary: ParentAssistantApiProviderBoundary,
    pub provider_route: ParentAssistantProviderRoute,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ParentAssistantRunCancelResult {
    pub schema_version: String,
    pub backend_state: ParentAssistantBackendState,
    pub thread_id: String,
    pub run_id: String,
    pub cancel_state: ParentAssistantRunCancelState,
    pub run_state: ParentAssistantRunState,
    pub provider_state: ParentAssistantProviderState,
    pub unavailable_reason: Option<String>,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ParentAssistantActionConfirmResult {
    pub schema_version: String,
    pub backend_state: ParentAssistantBackendState,
    pub action_intent_id: String,
    pub preview_id: Option<String>,
    pub action_kind: ParentAssistantActionPreviewKind,
    pub confirm_state: ParentAssistantActionConfirmState,
    pub preview_required: bool,
    pub preview_satisfied: bool,
    pub raw_assistant_prose_accepted: bool,
    pub parent_confirmation_required: bool,
    pub parent_confirmation_recorded: bool,
    pub child_agent_validation_state: ParentAssistantChildAgentValidationState,
    pub source_refs: Vec<ParentEvidenceReference>,
    pub audit_reason: String,
    pub requires_controller_lease: bool,
    pub child_agent_contract_required: bool,
    pub enforcement_applied: bool,
    pub policy_written: bool,
    pub reason: String,
}
